src/SparqlEngineMain.cpp

```diff
@@ -25,28 +25,62 @@ using std::cerr;
 
 // Available options.
 struct option options[] = {
-    {"queryfile",    required_argument, NULL, 'q'},
-    {"interactive",  no_argument,       NULL, 'I'},
-    {"index",        required_argument, NULL, 'i'},
-    {"text",         no_argument,       NULL, 't'},
-    {"cost-factors", required_argument, NULL, 'c'},
-    {"on-disk-literals",  no_argument,       NULL, 'l'},
-    {"all-permutations",  no_argument,       NULL, 'a'},
-    {NULL, 0,                          NULL, 0}
+  {"index",            required_argument, NULL, 'i'},
+  {"text",             no_argument,       NULL, 't'},
+  {"queryfile",        required_argument, NULL, 'q'},
+  {"interactive",      no_argument,       NULL, 'I'},
+  {"cost-factors",     required_argument, NULL, 'c'},
+  {"on-disk-literals", no_argument,       NULL, 'l'},
+  {"all-permutations", no_argument,       NULL, 'a'},
+  {"unopt-optional",   no_argument,       NULL, 'u'},
+  {"help",             no_argument,       NULL, 'h'},
+  {NULL,               0,                 NULL, 0}
 };
 
-void processQuery(QueryExecutionContext& qec, const string& query);
+void processQuery(QueryExecutionContext& qec, const string& query,
+                  bool optimizeOptionals);
+void printUsage(char *execName);
+
+void printUsage(char *execName) {
+  std::ios coutState(nullptr);
+  coutState.copyfmt(cout);
+  cout << std::setfill(' ') << std::left;
+
+  cout << "Usage: " << execName << " -i <index> [OPTIONS]"
+       << endl << endl;
+  cout << "Options" << endl;
+  cout << "  " << std::setw(20) << "a, all-permutations" << std::setw(1)
+       << "    "
+       << "Load all six permuations of the index instead of only two." << endl;
+  cout << "  " << std::setw(20) << "c, cost-factors" << std::setw(1)
+       << "    "
+       << "Path to a file containing cost factors." << endl;
+  cout << "  " << std::setw(20) << "h, help" << std::setw(1) << "    "
+       << "Show this help and exit." << endl;
+  cout << "  " << std::setw(20) << "i, index" << std::setw(1) << "    "
+       << "The location of the index files." << endl;
+  cout << "  " << std::setw(20) << "I, interactive" << std::setw(1)
+       << "    " << "Use stdin to read the queries." << endl;
+  cout << "  " << std::setw(20) << "l, on-disk-literals" << std::setw(1)
+       << "    "
+       << "Indicates that the literals can be found on disk with the index."
+       << endl;
+  cout << "  " << std::setw(20) << "q, queryfile" << std::setw(1)
+       << "    "
+       << "Path to a file containing one query per line." << endl;
+  cout << "  " << std::setw(20) << "t, text" << std::setw(1) << "    "
+       << "Enables the usage of text." << endl;
+  cout << "  " << std::setw(20) << "u, unopt-optional" << std::setw(1) << "    "
+       << "Always place optional joins at the root of the query execution tree."
```

Comment:
So this means they are done last, right? Maybe this would be easier to understand

Reply:
Changed to 'Always execute optional joins last.'